README.md
# VsPsN
Perl Speaks Nonmem Command Line tool for VSCode

CHANGELOG.md
# Change Log

All notable changes to the "psn-vscode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import { window, commands, ExtensionContext, Uri, Disposable } from 'vscode';
import { showExcuteCmdBox, showVPCCmdBox, showBootStrapCmdBox, showSCMCmdBox} from './InputBox';
// import { multiStepInput } from './multiStepInput';
// import { quickOpen } from './quickOpen';

export function activate(context: ExtensionContext) {
	let disposables : Disposable[] = [];
	var disposable = commands.registerCommand('vspsn.execute', (uri:Uri) => {
		showExcuteCmdBox(uri);
	  });
	disposables.push(disposable);

	disposable = commands.registerCommand('vspsn.diagnostics.vpc', (uri:Uri) => {
		showVPCCmdBox(uri);
	  });
	disposables.push(disposable);

	disposable = commands.registerCommand('vspsn.diagnostics.bootstrap', (uri:Uri) => {
		showBootStrapCmdBox(uri);
	  });
	disposables.push(disposable);

	disposable = commands.registerCommand('vspsn.covariates.scm', (uri:Uri) => {
		showSCMCmdBox(uri);
	  });
	disposables.push(disposable);
	

	context.subscriptions.push(...disposables);
}


src/basicInput.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { window } from 'vscode';
import * as path from 'path';

/**
 * Shows a pick list using window.showQuickPick().
 */
export async function showQuickPick() {
	let i = 0;
	const result = await window.showQuickPick(['eins', 'zwei', 'drei'], {
		placeHolder: 'eins, zwei or drei',
		onDidSelectItem: item => window.showInformationMessage(`Focus ${++i}: ${item}`)
	});
	window.showInformationMessage(`Got: ${result}`);
}

interface FilePath{
	dir : String;
	name: String;
	ext: String;
}

/**
 * Shows an input box using window.showInputBox().
 */
async function getDirNameExtFromAcriveTextEditor(): Promise<FilePath | undefined> {
	const fpath = await window.activeTextEditor?.document.fileName;
	if (fpath === undefined) {
		window.showInformationMessage(`open a mod file`);
		return;
	}
	const fext = path.extname(fpath);
	const fname = path.basename(fpath, fext);
	const fdir = path.dirname(fpath);
	
	const result : FilePath= { 
		dir: fdir,
		name: fname,
		ext: fext
	};
	return result;
}

export async function showExcuteCmdBox() {
	const fp = await getDirNameExtFromAcriveTextEditor();
	if (fp === undefined){return;}

	if (fp.ext !== '.mod' ) {
		window.showInformationMessage(`${fp.dir}/${fp.name}${fp.ext} is not a mod file`);
		return;
	}

	const cmdList = [];
	cmdList.push('execute');
	cmdList.push([fp.name, fp.ext].join(''));
	const cmdstr = cmdList.join(' ');

	const result = await window.showInputBox({
		value: cmdstr,
		placeHolder: 'psn execute cmd',
	});

	if (result === undefined) {
		return;
	}
	const psnTerminal = window.createTerminal({name: 'psn terminal', cwd: fp.dir.toString()});
	psnTerminal.sendText(result);
	psnTerminal.show();
};

export async function showVPCCmdBox() {
	const fp = await getDirNameExtFromAcriveTextEditor();
	if (fp === undefined){return;}

	if (fp.ext !== '.mod' ) {
		window.showInformationMessage(`${fp.dir}/${fp.name}${fp.ext} is not a mod file`);
		return;
	}
	

	const cmdList = [];
	cmdList.push('vpc -samples=200 -auto_bin=auto');
	cmdList.push(`-dir=vpc_${fp.name}`);
	cmdList.push([fp.name, fp.ext].join(''));
	const cmdstr = cmdList.join(' ');

	const result = await window.showInputBox({
		value: cmdstr,
		placeHolder: 'psn execute cmd',
	});

	if (result === undefined) {
		return;
	}
	const psnTerminal = window.createTerminal({name: 'psn terminal', cwd: fp.dir.toString()});
	psnTerminal.sendText(result);
	psnTerminal.show();
};

export async function showBootStrapCmdBox() {
	const fp = await getDirNameExtFromAcriveTextEditor();
	if (fp === undefined){return;}

	if (fp.ext !== '.mod' ) {
		window.showInformationMessage(`${fp.dir}/${fp.name}${fp.ext} is not a mod file`);
		return;
	}
	

	const cmdList = [];
	cmdList.push('bootstrap -samples=50 -threads=4');
	cmdList.push(`-dir=bs_${fp.name}`);
	cmdList.push([fp.name, fp.ext].join(''));
	const cmdstr = cmdList.join(' ');

	const result = await window.showInputBox({
		value: cmdstr,
		placeHolder: 'psn execute cmd',
	});

	if (result === undefined) {
		return;
	}
	const psnTerminal = window.createTerminal({name: 'psn terminal', cwd: fp.dir.toString()});
	psnTerminal.sendText(result);
	psnTerminal.show();
};

export async function showSCMCmdBox() {
	const fp = await getDirNameExtFromAcriveTextEditor();
	if (fp === undefined){return;}

	if (fp.ext !== '.mod' ) {
		window.showInformationMessage(`${fp.dir}/${fp.name}${fp.ext} is not a mod file`);
		return;
	}
	

	const cmdList = [];
	cmdList.push(`scm -config_file=${fp.name}.scm`);
	cmdList.push(`-model=${fp.name}${fp.ext}`);
	const cmdstr = cmdList.join(' ');

	const result = await window.showInputBox({
		value: cmdstr,
		placeHolder: 'psn execute cmd',
	});

	if (result === undefined) {
		return;
	}
	const psnTerminal = window.createTerminal({name: 'psn terminal', cwd: fp.dir.toString()});
	psnTerminal.sendText(result);
	psnTerminal.show();
};

src/InputBox.ts
import { window, Uri, workspace, ExtensionContext, env} from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import * as https from 'https';

// interface FilePath{
// 	dir : String;
// 	name: String;
// 	ext: String;
// }

// function getDirNameExt(uri : Uri): FilePath {
// 	const fpath = uri.fsPath;
// 	const fext = path.extname(fpath);
// 	const fname = path.basename(fpath, fext);
// 	const fdir = path.dirname(fpath);
	
// 	return { 
// 		dir: fdir,
// 		name: fname,
// 		ext: fext
// 	};
// }

function openTerminal(cwd: string, cmd: string) {
	const psnTerminal = window.createTerminal({name: `PsN Terminal: ${cmd}`, cwd: cwd});
	psnTerminal.sendText(cmd);
	psnTerminal.show();
}

async function checkUri(uri: Uri|undefined): Promise<Uri | undefined>{
	const activeUri = await window.activeTextEditor?.document.uri;
	if (uri === undefined && activeUri !== undefined) {
		return activeUri;
	} else {
		return uri;
	}
} 

export async function showExcuteCmdBox(uri : Uri|undefined) {

	var uri = await checkUri(uri);
	if (uri === undefined) {return;}
	
	const defaultCmdList = [];
	defaultCmdList.push('execute');
	defaultCmdList.push(`${path.basename(uri.fsPath)}`);
	const defaultCmd = defaultCmdList.join(' ');

	const inputCmd = await window.showInputBox({
		value: defaultCmd,
		placeHolder: 'PsN Command',
		title: 'PsN Command'
	});

	if (inputCmd === undefined) {return;}
	openTerminal(path.dirname(uri.fsPath), inputCmd);
};

export async function showVPCCmdBox(uri : Uri|undefined) {	

	var uri = await checkUri(uri);
	if (uri === undefined) {return;}

	const defaultCmdList = [];
	defaultCmdList.push('vpc -samples=200 -auto_bin=auto');
	defaultCmdList.push(`-dir=vpc_${path.basename(uri.fsPath, path.extname(uri.fsPath))}`);
	defaultCmdList.push(`${path.basename(uri.fsPath)}`);
	const defaultCmd = defaultCmdList.join(' ');

	const inputCmd = await window.showInputBox({
		value: defaultCmd,
		placeHolder: 'psn execute cmd',
	});

	if (inputCmd === undefined) {return;}
	openTerminal(path.dirname(uri.fsPath), inputCmd);
};

export async function showBootStrapCmdBox(uri: Uri|undefined) {
	var uri = await checkUri(uri);
	if (uri === undefined) {return;}

	const defaultCmdList = [];
	defaultCmdList.push('bootstrap -samples=50 -threads=4');
	defaultCmdList.push(`-dir=bs_${path.basename(uri.fsPath, path.extname(uri.fsPath))}`);
	defaultCmdList.push(`${path.basename(uri.fsPath)}`);
	const defaultCmd = defaultCmdList.join(' ');

	const inputCmd = await window.showInputBox({
		value: defaultCmd,
		placeHolder: 'psn execute cmd',
	});

	if (inputCmd === undefined) {return;}
	openTerminal(path.dirname(uri.fsPath).toString(), inputCmd);
};

export async function showSCMCmdBox(uri : Uri | undefined) {
	var uri = await checkUri(uri);
	if (uri === undefined) {return;}

	const modFileName = path.basename(uri.fsPath, path.extname(uri.fsPath));
	const existScm = fs.existsSync(`${uri.fsPath.replace('.mod', '')}.scm`);
	if (!existScm) {
		const result = await window.showQuickPick(['yes', 'no'], {
			placeHolder: 'yes/no',
			title: 'Create .scm File'
		});
		
		if (result === 'yes') {
			const scmUrl = "https://raw.githubusercontent.com/UUPharmacometrics/PsN/master/doc/config_template_standard.scm";
			const createdSCMPath = uri.fsPath.replace('.mod', '.scm');

			https.get(scmUrl, (response) => {
				// Create a write stream to save the file
				const fileStream = fs.createWriteStream(createdSCMPath);
			  
				// Pipe the response to the write stream
				response.pipe(fileStream);
			  
				// When the download is complete, close the write stream
				response.on('end', () => {
				  fileStream.close();
				  window.showTextDocument(Uri.file(createdSCMPath));
				});
			  }).on('error', (error) => {
				console.error(`Error downloading file: ${error}`);
			  });

			window.showInformationMessage(`please edit ${modFileName}.scm. and you can read a document about scm module from the link`, 'Open Link')
			.then(selection => {
				if (selection === 'Open Link') {
					env.openExternal(Uri.parse('https://github.com/UUPharmacometrics/PsN/releases/latest/download/scm_userguide.pdf'));
				}
			});
		} else {
			window.showInformationMessage("you can get a scm file from the link", 'Open Link')
			.then(selection => {
				if (selection === 'Open Link') {
					env.openExternal(Uri.parse('https://github.com/UUPharmacometrics/PsN/tree/master/doc'));
				}
			});
		} 
		
		return;
	}


	const defaultCmdList = [];
	defaultCmdList.push(`scm -config_file=${modFileName}.scm`);
	defaultCmdList.push(`-model=${path.basename(uri.fsPath)}`);
	const defaultCmd = defaultCmdList.join(' ');

	const inputCmd = await window.showInputBox({
		value: defaultCmd,
		placeHolder: 'psn execute cmd',
	});

	if (inputCmd === undefined) {return;}
	openTerminal(path.dirname(uri.fsPath), inputCmd);
};